### PR TITLE
Feature/radio checkout size

### PR DESCRIPTION
**Pull request by rottitime**
Closes: https://github.com/govuk-react/govuk-react/issues/597
Closes: https://github.com/i-dot-ai/govuk-react/issues/1

**Checklist**:

* [x] Documentation
* [x] Tests
* [x] Ready to be merged

### CODE REVIEW

**Comment by rottitime** [comment]
Hi, any updates on this? @penx

File: .gitignore [hunk 1]
```diff
@@ -46,4 +47,5 @@ govuk-react-*.tgz
 !.yarn/patches
 !.yarn/plugins
 !.yarn/releases
-!.yarn/versions
\ No newline at end of file
+!.yarn/versions
+yarn.lock
```
Comment (penx) [inline .gitignore:51]: please can you revert the changes to this file?

File: components/checkbox/src/index.tsx [hunk 2]
```diff
@@ -9,12 +9,14 @@ import styled from 'styled-components';
 import HintText from '@govuk-react/hint-text';
 import { FOCUS_COLOUR } from 'govuk-colours';
 import { BORDER_WIDTH, BORDER_WIDTH_FORM_ELEMENT, FOCUS_WIDTH, SPACING_POINTS } from '@govuk-react/constants';
-import { spacing, typography } from '@govuk-react/lib';
+import { spacing, typography, WithWhiteSpaceProps } from '@govuk-react/lib';
+import type { Size } from '@govuk-react/lib';
```
Comment (penx) [inline components/checkbox/src/index.tsx:13]: What's the motivation for making this a generic abstraction called Size?

Are there any drawbacks to explicitly listing the possible values in each component that has a size prop?

Comment (penx) [reply components/checkbox/src/index.tsx:13]: If it's an abstraction, is there a name we could use that is more specific to its use, such as `ComponentSizeVariant`?

File: components/checkbox/src/index.tsx [hunk 3]
```diff
@@ -112,9 +133,10 @@ Checkbox.displayName = 'Checkbox';
 Checkbox.defaultProps = {
   hint: undefined,
   className: undefined,
+  size: 'medium',
 };
 
-export interface CheckboxProps extends React.InputHTMLAttributes<HTMLInputElement> {
+export interface CheckboxProps extends Omit<React.InputHTMLAttributes<HTMLInputElement>, 'size'> {
```
Comment (penx) [inline components/checkbox/src/index.tsx:139]: Are there any downsides to preventing the user from accessing the html input size prop?

Would it be better to use a name for our size prop that didn't clash, such as `sizeVariant`?

File: components/checkbox/src/stories.tsx [hunk 4]
```diff
@@ -18,10 +18,17 @@ Default.args = {
...
+CheckboxSmall.args = {
+  children: 'Small checkbox',
+  size: 'small',
+};
+
 export const CheckboxDisabled = Template.bind({});
 CheckboxDisabled.args = {
   children: 'Disabled checkbox option',
   disabled: true,
+  size: 100,
```
Comment (penx) [inline components/checkbox/src/stories.tsx:31]: I'm not sure how `100` is valid here?

File: components/radio/src/__snapshots__/test.tsx.snap [hunk 5]
```diff
@@ -268,12 +271,15 @@ exports[`Radio disabled: disabled 1`] = `
...
   inline={false}
+  size="medium"
 >
   <src__Label
     inline={false}
+    size="medium"
   >
     <label
       className="c0"
+      size="medium"
```
Comment (penx) [inline components/radio/src/__snapshots__/test.tsx.snap:282]: do we want the size prop to be sent to the label element?

**Comment by penx** [comment]
Thanks for the PR @rottitime! Sorry I've been busy but have left some feedback and should have some time to look again tomorrow.

File: .gitignore (hunk 1 above)
Comment (rottitime) [reply .gitignore]: thanks will do

File: components/checkbox/src/index.tsx (hunk 2 above)
Comment (rottitime) [reply components/checkbox/src/index.tsx:13]: @penx I noticed we use below in some components. As the sizes are shared it would be really good to start combining in one common place. It would also mean the sizes can be consistent. Is that ok?

I will update:
- name to be `ComponentSizeVariant`
- Values to match existing values: `XLARGE`, `LARGE`, `MEDIUM`, `SMALL`

```
`XLARGE`, `LARGE`, `MEDIUM`, `SMALL`, `XL`, `L`, `M`, `S`

File: components/checkbox/src/stories.tsx (hunk 4 above)
Comment (rottitime) [reply components/checkbox/src/stories.tsx]: Well spotted. This one was accidental

File: components/radio/src/__snapshots__/test.tsx.snap (hunk 5 above)
Comment (rottitime) [reply components/radio/src/__snapshots__/test.tsx.snap:282]: @penx the jest snapshots differ from what we have printed to screen. The real output is as:

```
<label class="sc-gsBqHK dEjzGa"><input type="radio" class="sc-aKXVD eEwhHG"><span class="sc-kmATbt cLEvoP">Small radio</span></label>

**Comment by rottitime** [comment]
> Thanks for the PR @rottitime! Sorry I've been busy but have left some feedback and should have some time to look again tomorrow.

I made the updates and given feedback to others. Can you review again please? :)

File: .gitignore [hunk 6]
```diff
@@ -1,5 +1,6 @@
 # dependencies
 /node_modules
+package-lock.json
```
Comment (penx) [inline .gitignore:3]: I see gitignore as containing files that, under normal usage, could be on disk but we would want to ignore.

package-lock.json is not something users should have under normal usage, and is a sign that something has gone wrong (i.e. they've used npm instead of yarn)

I would prefer this was not added to gitignore

**Comment by penx** [comment]
This looks great, thanks for taking the time to submit.

One minor comment on the gitignore, otherwise I think I'm happy with this approach.

@stevesims @Loque- any thoughts?

File: components/checkbox/src/index.tsx (hunk 2 above)
Comment (penx) [reply components/checkbox/src/index.tsx]: yes, agreed it would be good to make sure the strings we pass in to the sizeVariant props across components are consistent, so I agree with the abstraction for that

File: .gitignore (hunk 6 above)
Comment (rottitime) [reply .gitignore]: thanks I understand. This has been reverted back too

**Review by penx — APPROVED** [review]

**Comment by penx** [comment]
Vertical centering of the small checkbox at 320px looks off. Is this correct?

Screenshot 2023-03-03 at 12 43 19

**Comment by rottitime** [comment]
> Vertical centering of the small checkbox at 320px looks off. Is this correct?
> 
> Screenshot 2023-03-03 at 12 43 19

Thanks @penx I made the fontsize and spacing more consistent for both Radio and checkboxes with:
https://design-system.service.gov.uk/components/checkboxes/small/index.html

**Comment by rottitime** [comment]
@penx is there any further action from my side? I am assuming not. Thanks again for all your feedback

File: components/checkbox/src/stories.tsx [hunk 7]
```diff
@@ -18,6 +18,12 @@ Default.args = {
   children: 'Example',
 };
 
+export const CheckboxSmall = Template.bind({});
+CheckboxSmall.args = {
+  children: 'Small checkbox',
+  sizeVariant: 'SMALL',
```
Comment (penx) [inline components/checkbox/src/stories.tsx:24]: I think the prop value that is passed in should be lowercase, i.e. it should be usable as `<Checkbox sizeVariant="small" />`

Comment (rottitime) [reply components/checkbox/src/stories.tsx:24]: @penx I don't mind but this clashes with other parts of your code. All sizes so far have been uppercase.
Wouldn't it be great if we can retro fit the type for `sizeVariant` on all existing component. It will keep everything more manageable from a code point of view?

If you still need it changed to lowercase, I can still do so?

Existing doc:
Screenshot 2023-03-08 at 20 46 47

Comment (penx) [reply components/checkbox/src/stories.tsx:24]: Ah ok. Thanks. Let me have a think

Comment (rottitime) [reply components/checkbox/src/stories.tsx:24]: > Ah ok. Thanks. Let me have a think

@penx do you have an update. 

I think we should keep the size names consistent with the rest of your code. If we can start making it TypeSafe for now then we can change it much easier later on. I am happy to go with whatever you decide

Comment (penx) [reply components/checkbox/src/stories.tsx:24]: sorry just to let you know I haven't forgotten about this but have had a busy few weeks - will hopefully find some time tomorrow to review

Comment (penx) [reply components/checkbox/src/stories.tsx:24]: ok great, confirmed Heading and Caption currently have a size prop that accepts a number or an uppercase string in this set.

We should be consistent here, and if we make a change it should be done across the library.

We're breaking from convention here and using `sizeVariant` instead of `size` so as not to conflict with an existing html attribute.

So... all good I think!

Are you hoping to see this in a release soon?

Comment (rottitime) [reply components/checkbox/src/stories.tsx:24]: > Are you hoping to see this in a release soon?

there was a urgency but no so much anymore. so approve and merge whenever you can :)
